Homework 4 mvvm

Voodstr opened the pull request:
Не реализована пагинация и работа с загрузкой изображений (GLIDE). 28.05 сделаю

Commit:
Update README.md

Commit:
Обновил зависимости для работы с ViewModel и LiveData.

Commit:
еду домой

Commit:
livedata+viewmodel added.
clicklisteners doesn't work

Commit:
Merge remote-tracking branch 'origin/homework_4_mvvm' into homework_4_mvvm

# Conflicts:
#	app/src/main/java/ru/voodster/otuslesson/film/FilmAdapter.kt
#	app/src/main/java/ru/voodster/otuslesson/film/FilmListFragment.kt
#	build.gradle

Commit:
film clicks works

Commit:
Room kinda works

Commit:
favorite list works

Commit:
diffUtil update for all adapters
Dao Update

Commit:
Добавлена своя вариация Пагинации из БД

Commit:
Подгрузка списка избранно с пагинацией. Добавлено

Commit:
Подгрузка списка избранно с пагинацией. Bug fix

Commit:
Создано локальное Api. Все переделано под его работу

baidala commented on .idea/codeStyles/Project.xml:
```diff
@@ -1,22 +1,6 @@
 <component name="ProjectCodeStyleConfiguration">
```
все файлы .idea/  лучше добавить в gitignore

Commit:
Создано локальное Api. Все переделано под его работу

Commit:
Рабочая версия, но есть проблема: иногда в списке кнопка не запускает отрисовку "сердечка"

Commit:
Подгрузка из БД при отсутствии интернета

baidala commented on app/src/main/java/ru/voodster/otuslesson/App.kt:
```diff
@@ -0,0 +1,72 @@
+
+        Log.d(TAG,"initRetrofit")
+
+        val okHttpClient = OkHttpClient.Builder()
```
рекомендовал бы добавить еще  htttpInterceptor  чтоб в дебаге выводить в лог тело запросов
```

implementation "com.squareup.okhttp3:logging-interceptor:4.9.0"
-----------
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.logging.HttpLoggingInterceptor.Level.BODY
import okhttp3.logging.HttpLoggingInterceptor.Level.NONE

fun provideLoggingInterceptor() =
            HttpLoggingInterceptor().apply { level = if (BuildConfig.DEBUG) BODY else NONE }
*****
val okHttpClient = OkHttpClient.Builder()
            .addInterceptor(provideLoggingInterceptor())
            .build()


```

baidala commented on app/src/main/java/ru/voodster/otuslesson/App.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    companion object{
+        const val BASE_URL = "http://10.0.2.2/"
```
это какой то локальный адресс? с какого сайтв вы берете данные про фильмы?
`okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: failed to connect to /10.0.2.2 (port 80) from /192.168.50.162 (port 45976) after 10000ms`

Voodstr replied:
Свой http сервер запускаю локально. Могу поднять на яндекс облаке в теории

baidala commented on app/src/main/java/ru/voodster/otuslesson/about/AboutFragment.kt:
```diff
@@ -2,18 +2,24 @@ package ru.voodster.otuslesson.about
+import ru.voodster.otuslesson.db.FilmModel
 
-class AboutFragment(private val filmID:Int) :Fragment() {
+class AboutFragment(private val film: FilmModel) :Fragment() {
```
у каждого фрагмента должен быть пустой дефолтный конструктор или при пересоздании/востановлении будут креши. и лучше добавить отдельный метод newInstanse(film: FilmModel) для аргументов. и в onViewCreated доставить эти данные из arguments
```
class AboutFragment :Fragment() {
private lateinit var film: FilmModel
    
    companion object{
        const val TAG = "AboutFragment"
        private const val FILM_DATA = "FILM_DATA"
        fun newInstance(film: FilmModel): AboutFragment {
            val args  = Bundle()
            args.putParcelable(FILM_DATA, film)
            val fragment = AboutFragment()
            fragment.arguments = args
            return fragment
        }
    }
```

baidala replied:
другой вариант записи
```
***
fun newInstance(film: FilmModel): AboutFragment {
           return AboutFragment().apply {
                arguments = Bundle().apply {
                    putParcelable(FILM_DATA, film)
                }
            }
        }

```

baidala commented on app/src/main/java/ru/voodster/otuslesson/about/AboutFragment.kt:
```diff
@@ -49,11 +56,14 @@ class AboutFragment(private val filmID:Int) :Fragment() {
+            it.findViewById<TextView>(R.id.aboutDescriptionTv).text = film.description.plus("\n")
+            Glide
+                .with(this)
+                .load("http://developer.alexanderklimov.ru/android/images/android_cat.jpg")
```
в aboutImg нужно  загружать картинку из FilmModel.img.  и рекомендую использовать  .placeholder(**)  и поставить какую то неброскую/однотонную картинку  пока грузится  постер

baidala commented on app/src/main/java/ru/voodster/otuslesson/db/Db.kt:
```diff
@@ -0,0 +1,277 @@
+                    App.instance!!.applicationContext,
+                    FilmsRoomDatabase::class.java, "db-name.db"
+                )
+                    /*.allowMainThreadQueries()
```
.allowMainThreadQueries()  - лучше не использовать. сразу приучайтесь что все запросы в сеть/базу  должны иддти в фоне

baidala commented on app/src/main/java/ru/voodster/otuslesson/api/FilmsApi.kt:
```diff
@@ -0,0 +1,15 @@
+
+
+    @GET("/get")
+    fun getInitial():Call<List<FilmModel>>
```
нужно разделить FilmModel - вы можете использовать эту модель в базе данных и на UI уровне (в больших проектах есть модель для базы. модель для rest запросов. и общая модель для UI к которой все сводится). а для сетевых запросов должна быть отдельная модель  для того чтоб эти два источника данных (база и сеть) были максимально независимы друг от друга.  т.к. сейчас - если изменится ответ сервера, вам  нужно переделыыать в двух местах. и при получении данных от сервера вы будете конвертировать в  FilmModel  и сохранять в базу. для этого используется класс -маппер.  пример базового маппера  
https://gist.github.com/baidala/6867528cf6d46a7b10d44370f9cccf27

baidala commented on app/src/main/java/ru/voodster/otuslesson/db/FilmModel.kt:
```diff
@@ -0,0 +1,32 @@
+import androidx.room.*
+import com.google.gson.annotations.SerializedName
+@Entity(tableName = "films_table")
+class FilmModel {
```
data class FilmModel  - и для других моделей тоже используйте  data class

baidala reviewed (changes requested):
в целом mvvm реализован правильно. fragment - viewmodel- interactor- dataSource. но нужно разделить FilmModel. указать правильные адресса. внести изменения в фрагменты.

Commit:
1. Glide работает везде.

Commit:
1. added http source file

Commit:
1. AboutFragment.kt default constructor added
2. Optimized cached files -  only one FilmList left
3. Glide works for every picture
4. Added FavoritesIDs to cache.
5. Added app/data/filmlist.json which is my http source for films

Commit:
Включил сервер https://84.252.141.104/

Commit:
Включил сервер https://84.252.141.104/

Commit:
ip change https://130.193.43.126/

Commit:
1. SingleLiveEvent
2. Подгрузка из базы при отсутствии интернета

Commit:
1. SingleLiveEvent
2. Подгрузка из базы при отсутствии интернета

Commit:
??????

Commit:
set all executors to queryExecutor

Commit:
fix fragment.onPause -> save

Commit:
works

Commit:
works on my own server

Commit:
works on my own server correctly

Commit:
server api fixes

Commit:
По нажатию на сердечко полностью удаляет элемент

baidala reviewed (approved):
можно мержить с веткой homework5

Commit:
Merge pull request #5 from Voodstr/homework5

Homework5
